### ISSUE TITLE
CI tests are not running because we don't have a broker

### ISSUE THREAD

**Issue by mobpakrd** [ryanhiebert/hirefire#39]
We’re setting up a django/celery integration based on the documentation here.

We’re having trouble running tests in CI, where we don’t have redis setup, as we have celery running in always_eager mode for testing purposes, so all tasks happen synchronously. This is causing the following error for all of our tests. Is there a recommended strategy for HireFire in tests? I'm not sure what, if any, code snippets would be useful here?

```
  File "/root/.local/lib/python2.7/site-packages/rest_framework/test.py", line 290, in post
    path, data=data, format=format, content_type=content_type, **extra)
  File "/root/.local/lib/python2.7/site-packages/rest_framework/test.py", line 212, in post
    return self.generic('POST', path, data, content_type, **extra)
  File "/root/.local/lib/python2.7/site-packages/django/test/client.py", line 416, in generic
    return self.request(**r)
  File "/root/.local/lib/python2.7/site-packages/rest_framework/test.py", line 279, in request
    return super(APIClient, self).request(**kwargs)
  File "/root/.local/lib/python2.7/site-packages/rest_framework/test.py", line 231, in request
    request = super(APIRequestFactory, self).request(**kwargs)
  File "/root/.local/lib/python2.7/site-packages/django/test/client.py", line 483, in request
    response = self.handler(environ)
  File "/root/.local/lib/python2.7/site-packages/django/test/client.py", line 131, in __call__
    self.load_middleware()
  File "/root/.local/lib/python2.7/site-packages/django/core/handlers/base.py", line 56, in load_middleware
    mw_class = import_string(middleware_path)
  File "/root/.local/lib/python2.7/site-packages/django/utils/module_loading.py", line 20, in import_string
    module = import_module(module_path)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/root/.local/lib/python2.7/site-packages/hirefire/contrib/django/middleware.py", line 34, in <module>
    class HireFireMiddleware(MiddlewareMixin):
  File "/root/.local/lib/python2.7/site-packages/hirefire/contrib/django/middleware.py", line 42, in HireFireMiddleware
    loaded_procs = load_procs(*PROCS)
  File "/root/.local/lib/python2.7/site-packages/hirefire/procs/__init__.py", line 53, in load_procs
    proc = load_proc(obj)
  File "/root/.local/lib/python2.7/site-packages/hirefire/procs/__init__.py", line 34, in load_proc
    proc = proc()
  File "/root/.local/lib/python2.7/site-packages/hirefire/procs/celery.py", line 237, in __init__
    self.channel = self.connection.channel()
  File "/root/.local/lib/python2.7/site-packages/kombu/connection.py", line 242, in channel
    chan = self.transport.create_channel(self.connection)
  File "/root/.local/lib/python2.7/site-packages/kombu/connection.py", line 742, in connection
    self._connection = self._establish_connection()
  File "/root/.local/lib/python2.7/site-packages/kombu/connection.py", line 697, in _establish_connection
    conn = self.transport.establish_connection()
  File "/root/.local/lib/python2.7/site-packages/kombu/transport/virtual/__init__.py", line 809, in establish_connection
    self._avail_channels.append(self.create_channel(self))
  File "/root/.local/lib/python2.7/site-packages/kombu/transport/virtual/__init__.py", line 791, in create_channel
    channel = self.Channel(connection)
  File "/root/.local/lib/python2.7/site-packages/kombu/transport/redis.py", line 466, in __init__
    self._disconnect_pools()
  File "/root/.local/lib/python2.7/site-packages/kombu/transport/redis.py", line 484, in _disconnect_pools
    self._async_pool.disconnect()
  File "/root/.local/lib/python2.7/site-packages/redis/connection.py", line 921, in disconnect
    connection.disconnect()
  File "/root/.local/lib/python2.7/site-packages/kombu/transport/redis.py", line 854, in disconnect
    channel._on_connection_disconnect(self)
  File "/root/.local/lib/python2.7/site-packages/kombu/transport/redis.py", line 498, in _on_connection_disconnect
    raise get_redis_ConnectionError()
ConnectionError
```

**Comment by ryanhiebert** [comment]
I use Travis, and I make sure to have Redis and AMQP (my results backend and broker) set up in the test environment. That'd be my first recommendation to you: test with the same tools as production.

**Comment by keturn** [comment]
Unit tests need to have the same initial conditions in every test.  For PostgreSQL, [django manages the test database](https://docs.djangoproject.com/en/1.11/topics/testing/overview/#the-test-database) with [django.test.TestCase](https://docs.djangoproject.com/en/1.11/topics/testing/tools/#testcase).

If we were to incorporate redis and RabbitMQ into our test environment, the test setup would need to reset those on every tearDown and they're in a clean state on setUp.

If our test environment is to no longer run celery tasks synchronously (or "eagerly"), then the test runner is going to have to know how to wait for any celery tasks launched to be complete before it declares the test complete, and it's going to have to handle any exceptions in tasks and report them as test failures.

Since the main code under test and the celery workers are now in different processes, using different database connections, none of this will work with the roll-back-the-transaction trick that django's `TestCase` uses to reset state at the end of this test. We'd have to fall back to the slower method their [other TestCase subclass](https://docs.djangoproject.com/en/1.11/topics/testing/tools/#transactiontestcase) uses.

And we'd have to make sure all that coordinates with Django's [parallel test runner](https://docs.djangoproject.com/en/1.11/ref/django-admin/#cmdoption-test-parallel), so the queues from different runners don't interfere with each other.

Granted, Celery's [warning about using eager mode](http://docs.celeryproject.org/en/v4.1.0/userguide/testing.html#tasks-and-unit-tests) is a very real concern:

> When testing with eager mode you are only testing an emulation of what happens in a worker, and there are many discrepancies between the emulation and what happens in reality.

So a test framework with the above features could be nice to have, but that's a very long way from being an out-of-the-box Django+Celery configuration.

Alternatively, hirefire's django integration could refrain from making a new connection to the celery broker on every request (even for requests and views that don't have any impact on hirefire). That seems like it would probably be easier.

**Comment by ryanhiebert** [comment]
I'm not opposed to trying to get some better characteristics as far as this goes. The procs are created at load time, but I'm not opposed to having them load up the channel lazily, instead of on initialization.

**Comment by ryanhiebert** [comment]
Is this addressed by #43? At least for Celery?

**Comment by ryanhiebert** [comment]
Cool. Closing this one then. @mobpakrd : care to test out the latest version, and see if it fixes your issue?
